### PR TITLE
osd: fix require_min_compat_client handling for msr rules

### PR DESCRIPTION
**Pull request by athanatos**
This PR fixes 3 problems:
1. check require_min_compat_client when creating a crush rule while creating a pool with an ec profile
2. consider a crushmap to require CRUSH_MSR if any rules are msr rules, even if not attached to a pool
3. use CEPH_FEATUREMASK_CRUSH_MSR rather than CEPH_FEATURE_CRUSH_MSR

See commit messages for details.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
jenkins test make check

**Review by ljflores — APPROVED** [review]
LGTM. I tested this using the following steps:

```
$ OSD=7 ../src/vstart.sh --debug --new -x --localhost --bluestore
$ ./bin/ceph osd pool create ec86-pool erasure ec86-profile
Error EINVAL: new crush map requires client version squid but require_min_compat_client is luminous
$ ./bin/ceph osd set-require-min-compat-client reef
set require_min_compat_client to reef
$ ./bin/ceph osd pool create ec86-pool erasure ec86-profile
Error EINVAL: new crush map requires client version squid but require_min_compat_client is reef
$ ./bin/ceph osd setcrushmap -i ~/bz_2302230/crush.map.bin // set it with an old crush map with an MSR rule
Error EINVAL: new crush map requires client version squid but require_min_compat_client is reef
$ ./bin/ceph osd set-require-min-compat-client squid
set require_min_compat_client to squid
$ ./bin/ceph osd pool create ec86-pool erasure ec86-profile
pool 'ec86-pool' created
$ ./bin/ceph osd set-require-min-compat-client reef
Error EPERM: osdmap current utilizes features that require squid; cannot set require_min_compat_client below that to reef
$ ./bin/ceph osd getcrushmap > crush.map.bin
16
$ ./bin/ceph osd setcrushmap -i crush.map.bin
17
```

**Comment by ljflores** [comment]
@athanatos so far I found one type of failure related to this PR. I'm still reviewing the upgrade suite, so full results pending.

```
/a/yuriw-2024-08-28_23:20:36-rados-wip-yuri4-testing-2024-08-28-1359-distro-default-smithi/7879335
/a/yuriw-2024-08-28_23:20:36-rados-wip-yuri4-testing-2024-08-28-1359-distro-default-smithi/7879444
/a/yuriw-2024-08-28_23:20:36-rados-wip-yuri4-testing-2024-08-28-1359-distro-default-smithi/7879549
2024-08-29T05:22:39.061 INFO:teuthology.orchestra.run.smithi070.stderr:noscrub is set
2024-08-29T05:22:39.076 INFO:tasks.rados:starting run 0 out of 1
2024-08-29T05:22:39.076 INFO:tasks.ceph.ceph_manager.ceph:creating pool_name unique_pool_0
2024-08-29T05:22:39.076 DEBUG:teuthology.orchestra.run.smithi070:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd pool create unique_pool_0 16 16 erasure jerasure21profile
2024-08-29T05:22:39.094 DEBUG:teuthology.orchestra.run.smithi096:> sudo adjust-ulimits ceph-objectstore-tool --err-to-stderr --no-mon-config '--log-file=/var/log/ceph/objectstore_tool.$pid.log' --data-path /var/lib/ceph/osd/ceph-6 --journal-path /var/lib/ceph/osd/ceph-6/journal --op export --pgid 2.6 --file /var/log/ceph/exp.2.6.6
2024-08-29T05:22:39.160 INFO:tasks.ceph.osd.2.smithi070.stderr:2024-08-29T05:22:39.158+0000 7fdfc6d87640 -1 received  signal: Hangup from /usr/bin/python3 /bin/daemon-helper kill ceph-osd -f --cluster ceph -i 2  (PID: 22460) UID: 0
2024-08-29T05:22:39.167 INFO:teuthology.orchestra.run.smithi070.stdout:ERROR: (22) Invalid argument
2024-08-29T05:22:39.167 INFO:teuthology.orchestra.run.smithi070.stdout:op_tracker tracking is not enabled now, so no ops are tracked currently, even those get stuck. Please enable "osd_enable_op_tracker", and the tracker will start to track new ops received afterwards.
2024-08-29T05:22:39.177 DEBUG:teuthology.orchestra.run.smithi070:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 30 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-osd.2.asok dump_ops_in_flight
2024-08-29T05:22:39.257 INFO:tasks.ceph.osd.8.smithi130.stderr:2024-08-29T05:22:39.256+0000 7f030f3e8640 -1 received  signal: Hangup from /usr/bin/python3 /bin/daemon-helper kill ceph-osd -f --cluster ceph -i 8  (PID: 22908) UID: 0
2024-08-29T05:22:39.324 INFO:teuthology.orchestra.run.smithi070.stdout:ERROR: (22) Invalid argument
2024-08-29T05:22:39.324 INFO:teuthology.orchestra.run.smithi070.stdout:op_tracker tracking is not enabled now, so no ops are tracked currently, even those get stuck. Please enable "osd_enable_op_tracker", and the tracker will start to track new ops received afterwards.
2024-08-29T05:22:39.333 DEBUG:teuthology.orchestra.run.smithi070:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 30 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-osd.2.asok dump_blocked_ops
2024-08-29T05:22:39.358 INFO:tasks.ceph.osd.7.smithi096.stderr:2024-08-29T05:22:39.357+0000 7fac60427640 -1 received  signal: Hangup from /usr/bin/python3 /bin/daemon-helper kill ceph-osd -f --cluster ceph -i 7  (PID: 21960) UID: 0
2024-08-29T05:22:39.422 INFO:teuthology.orchestra.run.smithi070.stderr:Error EINVAL: new crush map requires client version squid but require_min_compat_client is luminous
```

**Comment by athanatos** [comment]
Pushed hopefully a fix for ^ and started retest of a few of those jobs http://pulpito.ceph.com/sjust-2024-08-30_00:56:59-rados-wip-yuri4-testing-2024-08-28-1359-distro-default-smithi/

File: src/include/ceph_features.h [hunk 1]
```diff
@@ -218,7 +218,7 @@ DEFINE_CEPH_FEATURE_RETIRED(63, 1, RESERVED_BROKEN, LUMINOUS, QUINCY) // client-
 	 CEPH_FEATURE_OSD_CACHEPOOL |	    \
 	 CEPH_FEATURE_CRUSH_V2 |	    \
 	 CEPH_FEATURE_EXPORT_PEER |	    \
-	 CEPH_FEATURE_CRUSH_MSR |	    \
+	 CEPH_FEATUREMASK_CRUSH_MSR |	    \
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/include/ceph_features.h:221]: This needs to be backported immediately, so I'm optimizing for the simpler diff.  Longer term, I think we should deprecate usage of CEPH_FEATURE_ entirely and update existing usages where practical.

**Comment by athanatos** [comment]
@ljflores Looks like my commit above fixed that failure.

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues67783

**Comment by ljflores** [comment]
> Pushed hopefully a fix for ^ and started retest of a few of those jobs http://pulpito.ceph.com/sjust-2024-08-30_00:56:59-rados-wip-yuri4-testing-2024-08-28-1359-distro-default-smithi/

All green!

**Comment by ljflores** [comment]
jenkins test make check

**Pull request merged**